docs/ARCHITECTURE.md

```diff
@@ -84,6 +84,33 @@ Complex programs often require more than just simple functions to act as entry-p
     command.execute('goodbye')
     ```
 
+## Exceptions
+
+Large applications often have a correspondingly large number of failure points. For feature-level logic, these failures are often non-recoverable. The best we can do is show the user that something went wrong and maybe offer guidance on how to fix it.
+
+Because this is such a common pattern, shared error handling logic is defined by `ToolkitError` found [here](../src/shared/errors.ts). This class provides the basic structure for errors throughout the Toolkit.
+
+### Chaining
+
+Exceptions that occur deep in a call stack often do not contain enough context to correctly diagnose the problem. A stack trace is helpful, but only if the reader has access to both the source code and source map.
+
+By adding additional information as the exception bubbles up, we can create a better view of the program state when the problem occured. This is done via `ToolkitError.chain`. The `chain` function serves as a standard way to establish a clear cause-and-effect relationship for errors.
+
+### Handlers
+
+Any code paths exercised via `Commands` will have errors handled by `handleError` in [extensions.ts](../src/extension.ts). A better API for error handling across more than just commands will be added in a future PR.
+
+### Best Practices
+
+Implementations still need to adhere to some basic principles for this to work nicely:
+
+-   Do not catch errors unless something meaningful can be added.
+-   Do not swallow errors unless the functionality is non-critical to the feature.
+-   Do not directly show the user an error message for failed actions. Use `ToolkitError` instead.
+-   Use `chain` when re-throwing errors with additional information.
```

Comment:
The alternative is to instantiate the error directly:
```ts
throw new ToolkitError('Error!', { cause: err })
```

Which is the same as the examples given in https://github.com/tc39/proposal-error-cause
Can't think of any other languages that have utility functions for this stuff as each language does it slightly differently so I'll probably just get rid of `chain`.

Reply:
`append()`?

Reply:
Still kind of abstract, like what are we appending to?
It could be its own function but then you get discoverability issues.

I mainly wrote `chain` just because I didn't like typing out { cause: ... } all the time. Not sure if my minor annoyance is worth `chain`/`append`.

Reply:
`append` is starting to make more sense to me. Probably more intuitive too compared to `cause`.